### ISSUE TITLE
Parameter Missing Exception even though it is not missing

### ISSUE THREAD

**Issue by HoneyryderChuck** [rails/strong_parameters#162]
I'm implementing the strong_parameters gem on a rails 3.2 app (don't know the status of this in rails 4). I'm testing with rspec submission to a create action with invalid parameters. In this particular case I'm sending no parameters at all (an empty hash). Something like:

``` ruby
post :create, :user => {}
```

On the controller I'm requiring the user key using the strong parameters syntax:

``` ruby
params.require(:user).permit(:name, :password)
```

thing is, I'm getting the "ActionController::ParameterMissing: key not found: user" Exception message whenever I submit this empty hash. Granted, live the hash will always have something (csrf tag, utf8 val) but this seems kind of wrong. the user key is in fact there, so it should be checked whether the key exists and not if the value pointed to it is blank (nil or empty). Is it really like that?

**Comment by mjrk** [comment]
+1 scaffolds should be changed. When submitting empty forms - it results in an exception. This is really counter-intuitive. Newly generated scaffolds should kind of work.
The point with require() to be patched `nil?` and not `present?` is that it would not fix the issue for submitting empty forms.

IMHO the perfect solution: forms - even though all inputs are empty - should transmit the key with an empty hash and require should check for `nil?` then. But this would require too many changes. Too bad, at the end I cannot use strong parameters' `require()` anymore.

**Comment by MrTheWalrus** [comment]
+1 *something* here should be changed.  An empty form/parameter hash may or may not be a valid submission, but it's definitely not the kind of thing that strong params is *supposed* to be catching - that's what *validations* are for.

**Comment by hernan124** [comment]
+1

**Comment by askl56** [comment]
I agree. It demands needless complexity in your tests.

**Comment by fredngo** [comment]
Agreed. Ran into this most recently when a form had only radio inputs, and the user clicked submit without selecting an option. It turns out radio buttons don't create a parameter when they are empty. 

I agree with @MrTheWalrus that validations should handle when an option isn't selected, not strong parameters.

**Comment by fredngo** [comment]
I found myself implementing a hack similar to that described by @flatrocks all over the place, so I've replaced them all with this monkey patch. I know monkey patches are not in style anymore, but in this case it allows me to keep my controller code idiomatic. So for anyone else who might need something similar, here it is:

```ruby
module ActionController
  module ParametersExtensions

    # See https://github.com/rails/strong_parameters/issues/162
    def require key
      begin
        super key
      rescue ParameterMissing => e
        if self[key].nil?
          return Parameters.new
        else
          raise e
        end
      end
    end

  end

  class Parameters
    prepend ParametersExtensions
  end
end
```

**Comment by ryannealmes** [comment]
+1

**Comment by adilsoncarvalho** [comment]
I solved that by creating an instance of the parameters class and passed it to the method. Doing so it finds no issue with strong parameters `params.require(:user).permit(:email, :password)` blocks.

```ruby
post :create, ActionController::Parameters.new(user: { email: 'a@a.com', password: 'secret' })
```

I hope that it helps someone.
